plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaAccessControl.java

```diff
@@ -0,0 +1,978 @@
+                table -> buildQueryInputForSimpleResource(
+                        OpaQueryContext.fromSystemSecurityContext(context),
+                        "FilterTables",
+                        OpaQueryInputResource.builder()
+                                .table(TrinoTable.builder()
+                                        .catalogName(catalogName)
+                                        .schemaName(table.getSchemaName())
+                                        .tableName(table.getTableName())
+                                        .build())
+                                .build()));
+    }
+
+    @Override
+    public void checkCanShowColumns(SystemSecurityContext context, CatalogSchemaTableName table)
+    {
+        opaHighLevelClient.queryAndEnforce(
+                OpaQueryContext.fromSystemSecurityContext(context),
+                "ShowColumns",
+                AccessDeniedException::denyShowColumns,
+                table);
+    }
+
+    @Override
+    public Map<SchemaTableName, Set<String>> filterColumns(SystemSecurityContext context, String catalogName, Map<SchemaTableName, Set<String>> tableColumns)
+    {
+        ImmutableSet.Builder<TrinoColumn> allColumnsBuilder = ImmutableSet.builder();
+        Map<SchemaTableName, ImmutableSet.Builder<String>> resultBuilder = new HashMap<>();
+
+        for (Map.Entry<SchemaTableName, Set<String>> oneTableColumns : tableColumns.entrySet()) {
+            SchemaTableName schemaTableName = oneTableColumns.getKey();
+            resultBuilder.put(schemaTableName, ImmutableSet.builder());
+            for (String columnForTable : oneTableColumns.getValue()) {
+                allColumnsBuilder.add(new TrinoColumn(schemaTableName, columnForTable));
+            }
+        }
+
+        Set<TrinoColumn> filteredColumns = opaHighLevelClient.parallelFilterFromOpa(
+                allColumnsBuilder.build(),
+                tableColumn -> buildQueryInputForSimpleResource(
+                        OpaQueryContext.fromSystemSecurityContext(context),
+                        "FilterColumns",
+                        OpaQueryInputResource.builder()
+                                .table(TrinoTable.builder()
+                                        .catalogName(catalogName)
+                                        .schemaName(tableColumn.schemaTableName().getSchemaName())
+                                        .tableName(tableColumn.schemaTableName().getTableName())
+                                        .columns(ImmutableSet.of(tableColumn.columnName()))
+                                        .build())
+                                .build()));
+
+        for (TrinoColumn filteredColumn : filteredColumns) {
+            resultBuilder.get(filteredColumn.schemaTableName()).add(filteredColumn.columnName());
+        }
+
+        return resultBuilder.entrySet().stream().collect(toImmutableMap(Map.Entry::getKey, (mapEntry) -> mapEntry.getValue().build()));
+    }
```

Comment:
I think this is much simpler and doesn't require the helper TrinoColumn class:
```
    @Override
    public Map<SchemaTableName, Set<String>> filterColumns(SystemSecurityContext context, String catalogName, Map<SchemaTableName, Set<String>> tableColumns)
    {
        ImmutableSet.Builder<TrinoTable> allColumnsBuilder = ImmutableSet.builder();
        for (Map.Entry<SchemaTableName, Set<String>> entry : tableColumns.entrySet()) {
            SchemaTableName schemaTableName = entry.getKey();
            TrinoTable trinoTable = new TrinoTable(catalogName, schemaTableName.getSchemaName(), schemaTableName.getTableName());
            for (String columnName : entry.getValue()) {
                allColumnsBuilder.add(trinoTable.withColumns(ImmutableSet.of(columnName)));
            }
        }

        Set<TrinoTable> filteredColumns = opaHighLevelClient.parallelFilterFromOpa(
                allColumnsBuilder.build(),
                tableColumn -> buildQueryInputForSimpleResource(
                        OpaQueryContext.fromSystemSecurityContext(context),
                        "FilterColumns",
                        OpaQueryInputResource.builder().table(tableColumn).build()));

        ImmutableSetMultimap.Builder<SchemaTableName, String> results = ImmutableSetMultimap.builder();
        for (TrinoTable tableColumn : filteredColumns) {
            results.put(new SchemaTableName(tableColumn.schemaName(), tableColumn.tableName()), getOnlyElement(tableColumn.columns()));
        }
        return Multimaps.asMap(results.build());
    }
```

Reply:
Agreed this is simpler, I was trying to avoid building the internal `SchemaTableName` object myself but I imagine that is not a big deal.

A specific behaviour of the current implementation (and explicitly covered by my tests) is the case when a table has 0 columns after filtering. I have a test to assert that the keyset of the result is the same as the keyset of the input, tables with no permisisoned columns would just have an empty set in the map entry.

Is there any guidance as to whether this is needed? If Trino is OK with tables with 0 columns being missing from the map returned by `filterColumns` then your approach looks cleaner indeed

Reply:
I will for now modify the tests assuming that it is OK for the resulting map returned by these functions to be missing entries that have 0 items after filtering.

If that is in fact not OK please let me know and I will change it back